File: sdks/akta-docs/typescript/src/changelog.ts
import fs from "node:fs/promises";
import path from "node:path";
import os from "node:os";
import { ChangelogEntrySchema } from "./config/schema.js";
import type { AppendReport, ChangelogEntry } from "./types.js";

const ANCHOR = "<!-- akta-changelog-append -->";

export class ChangelogError extends Error {
  constructor(
    message: string,
    public readonly cause?: unknown,
  ) {
    super(message);
    this.name = "ChangelogError";
  }
}

function formatEntry(entry: ChangelogEntry, template: "compact" | "detailed"): string {
  const ts = entry.timestamp;
  const line = `- **${entry.change_type}** (${ts}) ${entry.summary}`;
  if (template === "detailed" && entry.details) {
    return `${line}\n  ${entry.details.split("\n").join("\n  ")}`;
  }
  return line;
}

export async function appendChangelogEntry(
  targetPath: string,
  rawPayload: unknown,
  options: { dryRun?: boolean; template?: "compact" | "detailed" } = {},
): Promise<AppendReport> {
  const parsed = ChangelogEntrySchema.safeParse(rawPayload);
  if (!parsed.success) {
    throw new ChangelogError(
      `Invalid changelog payload: ${parsed.error.flatten().formErrors.join("; ")}`,
      parsed.error,
    );
  }
  const entry = parsed.data;
  const template = options.template ?? "compact";
  const block = `\n${formatEntry(entry, template)}\n`;

  let existing: string | undefined;
  try {
    existing = await fs.readFile(targetPath, "utf8");
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code !== "ENOENT") throw e;
  }

  let out: string;
  let created = false;
  if (existing === undefined) {
    created = true;
    out = `---
doc_id: ${entry.doc_id}
doc_type: changelog
date: ${entry.timestamp.slice(0, 10)}
---

# Changelog

${ANCHOR}
${block}`;
  } else if (existing.includes(ANCHOR)) {
    out = existing.replace(ANCHOR, `${ANCHOR}${block}`);
  } else {
    const sep = existing.endsWith("\n") ? "\n" : "\n\n";
    out = `${existing}${sep}${block.trimStart()}\n`;
  }

  if (options.dryRun) {
    return {
      target: targetPath,
      bytes_written: Buffer.byteLength(out, "utf8"),
      created,
    };
  }

  const dir = path.dirname(targetPath);
  await fs.mkdir(dir, { recursive: true });
  const tmp = path.join(
    os.tmpdir(),
    `akta-changelog-${Date.now()}-${Math.random().toString(36).slice(2)}.md`,
  );
  await fs.writeFile(tmp, out, "utf8");
  await fs.rename(tmp, targetPath);

  return {
    target: targetPath,
    bytes_written: Buffer.byteLength(out, "utf8"),
    created,
  };
}

File: sdks/akta-docs/python/src/akta_docs/changelog.py
from __future__ import annotations

import json
import tempfile
from pathlib import Path

from pydantic import ValidationError

from akta_docs.models import ChangelogEntry

ANCHOR = "<!-- akta-changelog-append -->"


class ChangelogError(Exception):
    pass


def format_entry(entry: ChangelogEntry, template: str) -> str:
    line = f"- **{entry.change_type}** ({entry.timestamp}) {entry.summary}"
    if template == "detailed" and entry.details:
        indented = "\n  ".join(entry.details.split("\n"))
        return f"{line}\n  {indented}"
    return line


def append_changelog_entry(
    target_path: str | Path,
    payload: dict | str | bytes,
    *,
    dry_run: bool = False,
    template: str = "compact",
) -> tuple[str, int, bool]:
    if isinstance(payload, (str, bytes)):
        try:
            data = json.loads(payload)
        except json.JSONDecodeError as e:
            raise ChangelogError(f"Invalid JSON: {e}") from e
    else:
        data = payload
    try:
        entry = ChangelogEntry.model_validate(data)
    except ValidationError as e:
        raise ChangelogError(f"Invalid changelog payload: {e}") from e

    block = f"\n{format_entry(entry, template)}\n"
    path = Path(target_path)

    if path.exists():
        existing = path.read_text(encoding="utf-8")
    else:
        existing = None

    if existing is None:
        created = True
        out = f"""---
doc_id: {entry.doc_id}
doc_type: changelog
date: {entry.timestamp[:10]}
---

# Changelog

{ANCHOR}
{block}"""
    elif ANCHOR in existing:
        created = False
        out = existing.replace(ANCHOR, f"{ANCHOR}{block}")
    else:
        created = False
        sep = "\n" if existing.endswith("\n") else "\n\n"
        out = f"{existing}{sep}{block.strip()}\n"

    nbytes = len(out.encode("utf-8"))
    if dry_run:
        return str(path), nbytes, created

    path.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile(
        mode="w",
        encoding="utf-8",
        delete=False,
        dir=path.parent,
        prefix=".akta-changelog-",
        suffix=".md",
    ) as f:
        f.write(out)
        tmp = Path(f.name)
    tmp.replace(path)
    return str(path), nbytes, created
